File: modules/agent/src/models/UsageAnalytics.py
from sqlalchemy import Column, Text, DateTime, BigInteger, Integer, ForeignKey, func
from models import Base


class UsageAnalytics(Base):
    """Token usage analytics for tracking spend by node/tool/model."""

    __tablename__ = "Usage_Analytics"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    tenant_id = Column(BigInteger, ForeignKey("Tenant.id"), nullable=False)
    user_id = Column(BigInteger, ForeignKey("User.id"), nullable=False)
    conversation_id = Column(BigInteger, ForeignKey("Conversation.id"), nullable=True)
    message_id = Column(BigInteger, ForeignKey("Conversation_Message.id"), nullable=True)

    input_tokens = Column(Integer, nullable=False, default=0)
    output_tokens = Column(Integer, nullable=False, default=0)
    total_tokens = Column(Integer, nullable=False, default=0)

    node_name = Column(Text, nullable=True)
    tool_name = Column(Text, nullable=True)
    model_name = Column(Text, nullable=True)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

File: modules/agent/internal/models/UsageAnalytics.go
package models

import "time"

type UsageAnalytics struct {
	ID             int64     `gorm:"primaryKey;autoIncrement" json:"id"`
	TenantID       int64     `gorm:"not null;index" json:"tenant_id"`
	UserID         int64     `gorm:"not null;index" json:"user_id"`
	ConversationID *int64    `gorm:"index" json:"conversation_id"`
	MessageID      *int64    `gorm:"index" json:"message_id"`
	InputTokens    int       `gorm:"not null;default:0" json:"input_tokens"`
	OutputTokens   int       `gorm:"not null;default:0" json:"output_tokens"`
	TotalTokens    int       `gorm:"not null;default:0" json:"total_tokens"`
	NodeName       *string   `json:"node_name"`
	ToolName       *string   `json:"tool_name"`
	ModelName      *string   `json:"model_name"`
	CreatedAt      time.Time `gorm:"autoCreateTime" json:"created_at"`
}

func (UsageAnalytics) TableName() string {
	return "Usage_Analytics"
}
